### ISSUE TITLE
Audio loop timing

### ISSUE THREAD

**Issue by SinisterRectus** [SinisterRectus/Discordia#116]
Despite the fact that audio packets are sent with a sequence and timestamp value, packet send rates seem to affect the rate at which audio is reproduced by the Discord client. For small variations of a few milliseconds, this difference is unnoticeable, but if the actual elapsed time is very different from the expected time, the audio is not faithfully reproduced.

Therefore, while streaming audio, you want the packets to be released on a timeline that reproduces the original audio as faithfully as possible. Unfortunately, you can't just sleep 20 ms for every 20 ms packet because there are variations in how long it takes to process a packet and because it's impossible to sleep for an exact amount of time. To compensate, Discordia tracks how much audio time *should have* elapsed and how much audio time *has actually* elapsed. As the stream progresses, the loop attempts to bring the actual time as close to the expected time as possible by speeding up or slowing down based on the difference. This is essentially the P part of a [PID controller](https://en.wikipedia.org/wiki/PID_controller).

It appears that in #115, the FFmpegProcess is taking longer to read than expected at the start of the stream. The loop detects this and sends packets out with minimal delay until the stream catches up to the original audio. The result is rushed audio on the receiving end.

Potential solutions:
1. Debug FFmpegProcess so that the reads are consistent.
2. Impose a minimum and maximum delay on packet sending. 
3. Use a packet queue to buffer out the inconsistencies.
4. Incorporate the I and D parts of the PID controller.
4. Abandon the total elapsed time concept and try to adjust the timing according to only the most recent packets.

**Comment by Andols0** [comment]
The ``stream:read(0)`` example does not change anything.
Changing this ``sleep(max(delay, 0.5 * FRAME_DURATION))`` makes it sound better, but it will be sped up for longer since it takes more time for the delay to be <0.

For the first loop, if it is FFmpeg's startup time that is the problem could it be possible to add a check to see if there actually is anything to send before it sets ``start=hrtime()`` and enters the loop?

**Comment by SinisterRectus** [comment]
I've discovered the slow-down where I suspected it was occuring. Filling the audio buffer from the FFmpeg stream is about an order of magnitude slower than reading from it.

```lua
'filling buffer'
dt: 2.49224 | dm: 307.128 | avg dt: 2.49224 | avg dm: 307.128
dt: 0.298327 | dm: 75.5947 | avg dt: 1.39528 | avg dm: 191.361
dt: 0.258635 | dm: 71.8447 | avg dt: 1.0164 | avg dm: 151.522
...
'filling buffer'
dt: 2.24577 | dm: 217.601 | avg dt: 0.431771 | avg dm: 69.0873
dt: 0.351461 | dm: 76.0947 | avg dt: 0.427755 | avg dm: 69.4377
dt: 0.308569 | dm: 72.3447 | avg dt: 0.42208 | avg dm: 69.5761
...
'filling buffer'
dt: 2.34532 | dm: 283.192 | avg dt: 0.398558 | avg dm: 63.5604
dt: 0.20966 | dm: 76.5947 | avg dt: 0.393587 | avg dm: 63.9034
dt: 0.211901 | dm: 72.8447 | avg dt: 0.388929 | avg dm: 64.1326
...
```

I never noticed this because it was only 2 to 3 ms on my machine, but I suppose if you're running on an RPI then it's going to be more significant.

The reason why `stream:read(0)` was not effective was because it was failing the `#buffer < bytes` check. You can change it to `#buffer <= bytes` but that only fixes things on the first fill.

I'll have to see if I can find a method for decoupling the filling of the buffer from the reading.

If you want to play around with a simpler test, you can use this:

```lua
local sw = discordia.Stopwatch()
local stream = discordia.class.classes.FFmpegProcess('oof.wav', 48000, 2)

coroutine.wrap(function()
  for _ = 1, 10 do
    sw:reset()
    stream:read(1920)
    print(sw.milliseconds)
  end
  stream:close()
end)()
```

**Comment by SinisterRectus** [comment]
Can someone with a slow machine please test the following (pull the most recent library code):

```lua
local channel = client:getChannel('') -- insert your voice channel ID here
local conn = channel:join()
conn:playFFmpeg('oof.wav')
local f = assert(io.popen('ffmpeg -i oof.wav -ar 48000 -ac 2 -f s16le pipe:1 -loglevel warning'))
conn:playPCM(f)
conn:close()
```

You can benchmark the whole loop if you want, but I'm specifically interested in the read times:
```lua
open()
local pcm = stream:read(pcm_len)
if not pcm then break end
close()
```

**Comment by Andols0** [comment]
After some testing, files are [here](https://github.com/Andols0/Discordia/tree/master/Logs).

I played a local oog file that was 18 min long in parallel on 2 different guilds. Started them a few seconds in between. 1 & 2 is the outputs from using playFFmpeg 3 & 4 is from using the io.popen method.

File 1 to 4 is a save from 
``` lua
open()
local pcm = stream:read(pcm_len)
if not pcm then break end
close()
```
The the file Delays.txt is the output fom another check i added, it checks if the calculated delay is less than 0, if so it logs the current time in seconds from when i did the first play command, the id of the playthrough 1-4, and the current sequence ``self._s``, and the calculated delay. After a playthrough is done it logged some statistics.
Looking at the first files you can see that playFFmpeg got a lower average read time compared to popen.
But looking at the Delay.txt file it is less stable.
 playFFmpeg methods got 60 and 47 larger delays where the calculated delay got below 0, compared to popen that only got 1 and 2 larger drops.

Looking at the large start time it seems to mainly be the first one playing that gets the higher time 
``!play oof.ogg`` 
```
2018-04-18 09:22:16 | [INFO]    | Voice : Received READY
popen	
dt: 506.222 | dm: 50.332 | avg dt: 506.222 | avg dm: 50.332
dt: 0.347552 | dm: 18.8447 | avg dt: 253.285 | avg dm: 34.5884
dt: 0.342135 | dm: 18.8447 | avg dt: 168.971 | avg dm: 29.3405
dt: 0.41578 | dm: 18.8447 | avg dt: 126.832 | avg dm: 26.7166
dt: 0.337187 | dm: 18.8447 | avg dt: 101.533 | avg dm: 25.1422
dt: 0.359062 | dm: 18.8447 | avg dt: 84.6706 | avg dm: 24.0926
etc...
FFMPEG
Guessed Channel Layout for Input Stream #0.0 : stereo
dt: 1.15443 | dm: 199.358 | avg dt: 7.25639 | avg dm: 21.4183
dt: 0.438593 | dm: 71.5947 | avg dt: 7.16548 | avg dm: 22.0873
dt: 0.444531 | dm: 67.8447 | avg dt: 7.07705 | avg dm: 22.6894
dt: 0.619322 | dm: 64.0947 | avg dt: 6.99318 | avg dm: 23.2271
dt: 0.44776 | dm: 60.3447 | avg dt: 6.90927 | avg dm: 23.703
```
Move them around
```
2018-04-18 09:25:41 | [INFO]    | Voice : Received READY
FFMPEG
Guessed Channel Layout for Input Stream #0.0 : stereo
dt: 481.908 | dm: 59.0322 | avg dt: 481.908 | avg dm: 59.0322
dt: 0.32875 | dm: 15.5947 | avg dt: 241.118 | avg dm: 37.3135
dt: 1.31369 | dm: 316.562 | avg dt: 161.183 | avg dm: 130.396
dt: 0.528176 | dm: 72.0947 | avg dt: 121.02 | avg dm: 115.821
...
popen
dt: 2.69077 | dm: 35.6895 | avg dt: 7.06189 | avg dm: 53.0051
dt: 0.427447 | dm: 18.8447 | avg dt: 6.97344 | avg dm: 52.5496
dt: 0.445312 | dm: 18.8447 | avg dt: 6.88754 | avg dm: 52.1061
dt: 0.412447 | dm: 18.8447 | avg dt: 6.80345 | avg dm: 51.6742
```

**Comment by SinisterRectus** [comment]
`FFmpegProcess:read` (which currently uses uv.spawn) occasionally reads large chunks from the spawned process (this is slow). Otherwise, it reads from a string buffer (this is fast). That explains the occasional jumps in time used.

`PCMStream:read` reads from the spawned process on every call, so the performance is consistent.

It's unfortunate that the average time is a bit longer for reading the popen'd stream, but I think that the consistency, simplicity, and reduced garbage generation outweighs this.

My best explanation for the "starting first" anomaly is that LuaJIT is compiling some code on the first read, which is reused on later reads. You can test this by calling `jit.off` to disable JIT compilation.

**Comment by SinisterRectus** [comment]
I've switched to using `popen` for process spawning with 800082269d4ec7acc3a272f035fb3c3f721866ee

**Comment by Andols0** [comment]
Turning Jit on or off does not make any difference (if i did it correctly).

Going back to the first thing that it speeds up.
After listening to the same 3 songs way to many times this is my theory.
Discord tries to keep a buffer of around 80-100 ms. And if the buffer gets larger then that it speeds up the playback to shrink it and if it gets to large it skips ahead to catch up.
I have also noticed the playback slow down, maybe it tries to build up the buffer? (can't reproduce at will)

A thing that I'm trying now is this pice of code
```lua
elapsed = elapsed + FRAME_DURATION
local delay = elapsed - (hrtime() - start) * MS_PER_NS
if delay<-100 then
   start = hrtime() - (elapsed+80)/MS_PER_NS
end
```
The idea is that if the delay gets to large, reset the start value but keep a delay of 80 so that it sends enough data to fill the buffer without it speeding up or clipping.

**Comment by SinisterRectus** [comment]
I planning to just add a buffer for outgoing packets. It should be relatively simple to implement. Something from 100 to 1000 ms should be sufficient.

**Comment by SinisterRectus** [comment]
I may also implement something that resets the start time if recoverable delays are encountered. It's better to have temporary silence than it is to rush the packets.

**Comment by SinisterRectus** [comment]
There also appears to be a bug with `io.popen` on Linux where the process seemingly randomly and prematurely reaches EOF. Maybe this is an oversight on my part and maybe it can still happen with `uv.spawn`, but I will be switching back to `uv.spawn`. If not to fix this issue, then at least to regain stdin and stderr handles. I will also re-write the read logic so that less time is spent waiting for reads.

**Comment by SinisterRectus** [comment]
I re-implemented `FFmpegProcess` using `uv.spawn`, this time with preemptive buffer filling. If anything, this should remove the relatively long wait at the start of the stream. It may improve time spent reading in the middle of the stream. See `7cccb30416b71fcce5789ab88299b6aee85c5cb0`.

**Comment by SinisterRectus** [comment]
Do let me know if this resolves the major delay and rushing issue at the start of the stream.

**Comment by Andols0** [comment]
It fixes the start of the stream, but after a few seconds it gets worse :(
[Log](https://github.com/Andols0/Discordia/blob/master/Logs/With%20fill.txt)

**Comment by SinisterRectus** [comment]
I tried. I have all sorts of ideas on how to make it better, but I can't spend much more time implementing them right now if this is an issue that only occurs on slow processors and only for FFmpeg playing. I'm going to focus on getting a release out and then I will optimize the audio code.

**Comment by Mehgugs** [comment]
Perhaps making discordia voice run in a separate process communicating with the main via ipc or similar would fix these issues, although that it perhaps out of scope for the library itself.

**Comment by SinisterRectus** [comment]
Due to these issues, I've removed `playFFmpeg`  with b10398fb3587f904a68c8661d7f3389124885046 indefinitely. Users can still push data from an FFmpeg process into Discordia via `playPCM`, but they have to do it manually.

I may restore this functionality when I figure out a stable method for implementing it.

**Comment by SinisterRectus** [comment]
Note that removing `playFFmpeg` does not necessarily resolve these timing issues. This just simplifies the audio code and features so that I can more easily focus on what needs fixing.

**Comment by SinisterRectus** [comment]
Apparently, @Andols0's FFmpegProcess is slow in the middle because a large Lua string is being manipulated. At the end, the string grows shorter as the stream runs dry, and the reading becomes fast. The string is large in the middle because of the preemptive buffer filling.

**Comment by SinisterRectus** [comment]
I made a compromise in 74e823d323b115ff47a64c76d9370c7c311f0569. The buffer is filled once on `FFmpegProcess` initialization (new code), and as necessary all times after that (old code). This is as much time as I'm willing to spend on this issue. We can revisit in the future if necessary.
